formatters/event_rate_limit/event_rate_limit.go

```diff
@@ -0,0 +1,135 @@
+package event_rate_limit
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sort"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/openconfig/gnmic/formatters"
+	"github.com/openconfig/gnmic/types"
+	"github.com/openconfig/gnmic/utils"
+)
+
+const (
+	processorType = "event-rate-limit"
+	loggingPrefix = "[" + processorType + "] "
+	defaultCacheSize = 1000
+	oneSecond int64 = int64(time.Second)
+)
+
+// RateLimit rate-limits the message to the given rate.
+type RateLimit struct {
+	// formatters.EventProcessor
+
+	PerSecondLimit float64 `mapstructure:"per-second,omitempty" json:"per-second,omitempty"`
+	CacheSize        int     `mapstructure:"cache-size,omitempty" json:"cache-size,omitempty"`
+	Debug          bool    `mapstructure:"debug,omitempty" json:"debug,omitempty"`
+
+	// eventIndex is an lru cache used to compare the events hash with known value.
+	// LRU cache seems like a good choice because we expect the rate-limiter to be
+	// most useful in burst scenarios.
+	// We need some form of control over the size of the cache to contain RAM usage
+	// so LRU is good in that respect also.
+	eventIndex *lru.Cache[string, int64]
+	logger     *log.Logger
+}
+
+func init() {
+	formatters.Register(processorType, func() formatters.EventProcessor {
+		return &RateLimit{
+			logger:     log.New(io.Discard, "", 0),
+		}
+	})
+}
+
+func (o *RateLimit) Init(cfg interface{}, opts ...formatters.Option) error {
+	err := formatters.DecodeConfig(cfg, o)
+	if err != nil {
+		return err
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	if o.CacheSize <= 0 {
+		o.logger.Printf("using default value for lru size %d", defaultCacheSize)
+		o.CacheSize = defaultCacheSize
+
+	}
+	if o.PerSecondLimit <= 0 {
+		return fmt.Errorf("provided limit is %f, must be greater than 0", o.PerSecondLimit)
+	}
+	if o.logger.Writer() != io.Discard {
+		b, err := json.Marshal(o)
+		if err != nil {
+			o.logger.Printf("initialized processor '%s': %+v", processorType, o)
+			return nil
+		}
+		o.logger.Printf("initialized processor '%s': %s", processorType, string(b))
+	}
+
+	o.eventIndex, err = lru.New[string, int64](o.CacheSize)
+	if err != nil {
+		return fmt.Errorf("failed to initialize cache: %w", err)
+	}
+	return nil
+}
+
+func (o *RateLimit) Apply(es ...*formatters.EventMsg) []*formatters.EventMsg {
+	validEs := []*formatters.EventMsg{}
```

Comment:
maybe worth allocating the capacity of the slice here: `validEs := make([]*formatters.EventMsg, 0, len(es))`

Reply:
ok, it's not an issue if there are nil values in the slice that is returned by the processor then ?
I can have a counter inside the filter loop, that increments and pushes accepted events in the slice, but that means that there will be nil values at the end of the slice.

Or should I re-slice it before returning it ?

Reply:
I went for allocating the whole slice, and then slicing it at the end, based on the accepted count.
I actually never looked at make with more than one int :), but I can revert to that if you prefer.

Reply:
When creating a slice you can specify both its length and capacity: `make([]any, len, cap)`. 
The capacity is defined as `the maximum number of elements that the underlying array can hold without reallocation.`
This allows appending elements to the slice up to the capacity without having to reallocate the underlying array. There will be no nil elements at the end of the array.
I think it's better to set the length to 0, the capacity to `len(es)` and use `append()`.

Reply:
done :)